cuqi/implicitprior/_restorator.py

```diff
@@ -0,0 +1,218 @@
+#%%
+from abc import ABC, abstractmethod
+from cuqi.distribution import Distribution
+import numpy as np
+    
+class RestorationPrior(Distribution):
+    """    
+    This class defines an implicit distribution associated with a restoration operator
+    (eg denoiser). They are several works relating restorations operators with
+    priors, see 
+        -Laumont et al. https://arxiv.org/abs/2103.04715
+        -Hu et al. https://openreview.net/pdf?id=x7d1qXEn1e
+    We cannot sample from this distribution, neither compute its logpdf. It allows
+    us to apply algorithms such as MYULA and PnPULA.
+    
+    Parameters
+    ---------- 
+    restorator : callable f(x, restoration_strength)
+        Function f that accepts input x to be restored and returns the
+        restored version of x and information about the restoration operation.
+        
+    restorator_strength : float
+        Strength of the restoration operation. In the case where the restorator is
+        a denoiser, this parameter might correspond to the noise level.
+        
+    restorator_kwargs : dictionary
+        Dictionary containing information about the restorator.
+    
+    potential : callable function
+        In the case, where the restoration operator is the proximal operator of
+        the function potential. This function is a mapping from the image domain
+        to the real set. 
+    """
+
+    def __init__(self, restorator, restoration_strength=None, restorator_kwargs
+                =None, potential=None, **kwargs):
+        if restorator_kwargs is None:
+            restorator_kwargs = {}
+        self.restorator = restorator 
+        self.restorator_kwargs = restorator_kwargs
+        self.restoration_strength = restoration_strength
+        self.potential = potential
+        super().__init__(**kwargs)
+
+    def restorate(self, x):
+        """This function allows us to denoise the input x and returns the
```

Comment:
Not necessarily denoise, but restorate? Could say "restorate (e.g. denoise)"

Reply:
"restorate" is not a correct English word, perhaps just use "restore".

Reply:
Thank you Jakob! That is completely right. We modified it in our new commit and now use the word "restore" instead of "denoise" and "restorate"